pkg/controller/pravegacluster/pravegacluster_controller.go

```diff
@@ -159,13 +159,34 @@ func (r *ReconcilePravegaCluster) run(p *pravegav1beta1.PravegaCluster) (err err
 }
 
 func (r *ReconcilePravegaCluster) reconcileFinalizers(p *pravegav1beta1.PravegaCluster) (err error) {
-	zkFinalizer := "cleanUpZookeeper"
-	if util.ContainsString(p.ObjectMeta.Finalizers, zkFinalizer) {
-		p.ObjectMeta.Finalizers = util.RemoveString(p.ObjectMeta.Finalizers, zkFinalizer)
-		if err = r.client.Update(context.TODO(), p); err != nil {
-			return fmt.Errorf("failed to remove Zk Finalizer from Pravega object (%s): %v", p.Name, err)
+	if p.DeletionTimestamp.IsZero() {
+		if !util.ContainsString(p.ObjectMeta.Finalizers, util.ZkFinalizer) {
+			p.ObjectMeta.Finalizers = append(p.ObjectMeta.Finalizers, util.ZkFinalizer)
+			if err = r.client.Update(context.TODO(), p); err != nil {
+				return fmt.Errorf("failed to add the finalizer (%s): %v", p.Name, err)
+			}
+		}
+	} else {
+		if util.ContainsString(p.ObjectMeta.Finalizers, util.ZkFinalizer) {
+			p.ObjectMeta.Finalizers = util.RemoveString(p.ObjectMeta.Finalizers, util.ZkFinalizer)
+			if err = r.client.Update(context.TODO(), p); err != nil {
+				return fmt.Errorf("failed to update Pravega object (%s): %v", p.Name, err)
+			}
+			if err = r.cleanUpZookeeperMeta(p); err != nil {
```

Comment:
is it not required to send the event in the case of cleanup of meta data fails

Reply:
done